src/server/db_slice.cc

```diff
@@ -610,11 +621,20 @@ tuple<PrimeIterator, ExpireIterator, bool> DbSlice::AddOrFind2(const Context& cn
       existing->second.Reset();
       events_.expired_keys++;
 
-      return make_tuple(existing, ExpireIterator{}, true);
+      return {.it = existing,
```

Comment:
Whats a horrible function!
I dont think I can realy review this..
Anyway I believe we should call here the preupdate and remove the memory accounting from here. 
This is a bug f.e if we use tiering we will not free storage or cancel the pending IO.
Maybe worth to simplify this function in a different PR and rebase it to your branch. WDYT?

Reply:
> Whats a horrible function!

You mean that a super core function of 150 LoC with like 10 early returns is not easy to review? :rofl: 

> Maybe worth to simplify this function in a different PR and rebase it to your branch. WDYT?

That sounds like a good idea, I'll try that

Reply:
Done, ptal